### PR TITLE
Op kernel type reduction infrastructure.

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Add infrastructure to support type reduction in Op kernel implementations.
Update Cast and IsInf CPU kernels to use it.

**Motivation and Context**
Binary size reduction.

### CODE REVIEW

File: include/onnxruntime/core/framework/data_types_internal.h [hunk 1]
```diff
@@ -341,25 +345,51 @@ class MLTypeCallDispatcherRet {
   }
 };
 
-// Version of the MLTypeDispatcher that has an input type which is passed through ('carried')
-// as the first type parameter in the call to Fn when dispatching.
-template <typename TCarried, template <typename, typename> class Fn, typename... Types>
-class MLTypeCallDispatcherWithCarriedType {
+// TODO consolidate this with the other MLTypeCallDispatcher classes
+// can add additional methods to cover their usages, but need to update call sites
+template <typename... Types>
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/data_types_internal.h:350]: Comment should explain what the difference is vs MLTypeCallDispatcher (move Fn to Invoke) and why we need this.

May also be work considering replacing MLTypeCallDispatcher with MLTypeCallDispatcher2, but that can be a separate PR. Not sure there's any reason why we need a variant that takes Fn at the class level.

File: include/onnxruntime/core/framework/data_types_internal.h [hunk 2]
```diff
@@ -341,25 +345,51 @@ class MLTypeCallDispatcherRet {
...
-    int results[] = {0, helper.template Invoke<Types>(Fn<TCarried, Types>(), std::forward<Args>(args)...)...};
-    ORT_UNUSED_PARAMETER(results);
-    ORT_ENFORCE(helper.called_ < 2, "Check for duplicate types in MLTypeCallDispatcher");
+
+    std::array<int, sizeof...(Types)>{
+        helper.template Invoke<Types>(Fn<Types>(), std::forward<Args>(args)...)...};
+
+    // avoid "unused parameter" warning for the case where Types is empty
+    std::array<int, sizeof...(Args)>{(ORT_UNUSED_PARAMETER(args), 0)...};
+
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/data_types_internal.h:369]: Does an empty Types list occur in valid usage? 

Depending on that answer, should there be a static_assert that it's not empty (if that's not valid), or and 'if' using mp11::mp_empty to handle that case more explicitly?

File: onnxruntime/core/providers/op_kernel_type_control.h [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+}  // namespace tags
+
+// holds specified Op kernel argument types
+// if types are defined, the data member 'value' should contain them in a type list
+// see the tags in onnxruntime::op_kernel_type_control::tags for intended uses
+template <typename Tag>
+struct OpKernelArgTypes {};
+
+// gives access to the enabled Op kernel argument types via the 'type' data member
+template <typename OpKernelArgTag>
```
Comment (skottmckay) [inline onnxruntime/core/providers/op_kernel_type_control.h:64]: would be good to explicitly call out that if there are no types in the list the result is no data member 'type'. 
or maybe the comment for OpKernelArgTypes should say 'type' instead of 'value'?

File: onnxruntime/core/providers/op_kernel_type_control.h [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+struct EnabledOpKernelArgTypes {
+ private:
+  template <typename T>
+  using GetTypeMember = typename T::type;
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/op_kernel_type_control.h:69]: would it be clearer to call it 'types'/HasTypes instead of 'type'/HasTypeMember given it's a list?

**Comment by skottmckay** [comment]
> 

Is this where the script should add lists?

It may need to be in the framework library so it can be applied to contrib ops and ops in other providers (as they tend to be separate libraries) as well.

---
Refers to: onnxruntime/core/providers/op_kernel_type_control_overrides.inc:11 in 882857a. [](commit_id = 882857a5671148e7bd9011f671a31c511e8e9ed6, deletion_comment = False)

File: onnxruntime/test/providers/op_kernel_type_control_test.cc [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  static constexpr bool value =
+      (boost::mp11::mp_size<A>::value == boost::mp11::mp_size<B>::value) &&
+      (boost::mp11::mp_size<ABIntersection>::value == boost::mp11::mp_size<A>::value);
+};
+}  // namespace
+
+}  // namespace test
+
+// specify supported and allowed
+namespace op_kernel_type_control {
```
Comment (skottmckay) [inline onnxruntime/test/providers/op_kernel_type_control_test.cc:33]: Is it possible to test the usage of the global list here?

File: onnxruntime/core/providers/cpu/tensor/cast_op.cc [hunk 6]
```diff
@@ -1,193 +1,231 @@
...
 
 // handle floating point input separately
 template <typename SrcType>
 typename std::enable_if<std::is_floating_point<SrcType>::value, void>::type
-CastToStringData(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  const int64_t len = shape.Size();
-  const auto input_data = in.DataAsSpan<SrcType>();
-  auto output_data = out.MutableDataAsSpan<std::string>();
-
-  for (int i = 0; i < len; ++i) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/tensor/cast_op.cc:91]: Wondering if there's any potential perf impact from moving the 'for' loop away from the casting logic. Possibly not, but may need to check.

File: onnxruntime/core/providers/cpu/tensor/cast_op.cc [hunk 7]
```diff
@@ -1,193 +1,231 @@
...
-    }
-  } else {
-#ifdef ORT_NO_RTTI
-    ORT_THROW("Unsupported type in cast op");
-#else
-    ORT_THROW("Unsupported type in cast op: from String to ", typeid(DstType).name());
-#endif
+typename std::enable_if<std::is_integral<DstType>::value && std::is_unsigned<DstType>::value, void>::type
+CastFromString(const std::string& input, DstType& output) {
+  static_assert(sizeof(DstType) <= sizeof(unsigned long long),
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/tensor/cast_op.cc:106]: nit: would using uint64_t be clearer than 'unsigned long long'?

File: onnxruntime/core/providers/cpu/tensor/cast_op.cc [hunk 8]
```diff
@@ -1,193 +1,231 @@
...
+    std::string);
+}  // namespace op_kernel_type_control
+
 namespace {
-template <typename SrcType, typename DstType>
-inline void CastData(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  ptrdiff_t shape_size = gsl::narrow<ptrdiff_t>(shape.Size());
-  auto in_vector = ConstEigenVectorMap<SrcType>(in.Data<SrcType>(), shape_size);
-  auto output_vector = EigenVectorMap<DstType>(out.MutableData<DstType>(), shape_size);
-  output_vector = in_vector.template cast<DstType>();
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/tensor/cast_op.cc:32]: Any potential perf impact from not using Eigen for this? Not sure if Eigen has any optimizations for this operation that we'd be losing.

File: onnxruntime/core/providers/op_kernel_type_control_overrides.inc [hunk 9]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace onnxruntime {
+namespace op_kernel_type_control {
+
+// specify allowed types here
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/op_kernel_type_control_overrides.inc:8]: Would be good to have one example entry for a global list, and a per operator list.

May also be good to add tags for begin/end insertion points so the reduction script knows where to put things. Usage would be it replaces anything between the begin and end markers with new content (allows it to be run multiple times), and leaves anything outside the markers alone.

File: include/onnxruntime/core/framework/data_types_internal.h (hunk 2 above)
Comment (edgchen1) [reply include/onnxruntime/core/framework/data_types_internal.h:369]: i ran into it with Cast with a single type. there, we have two dispatchers: one with type list SrcTypes and the other with type list (DstTypes - given SrcType). when SrcTypes == DstTypes and they have a single element, the second dispatcher will have an empty type list. i think it's fine for this to be valid. calling Invoke() will fail though.
not sure whether `if(mp_empty<TypeList<Types...>>::value)` would make much difference here other than being notational - the problem is a compiler warning about the args pack. i think both branches would still be "seen" by the compiler. with C++17's if constexpr that could be more explicit though.

---
In reply to: [564866954](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564866954) [](ancestors = 564866954)

File: include/onnxruntime/core/framework/data_types_internal.h (hunk 1 above)
Comment (edgchen1) [reply include/onnxruntime/core/framework/data_types_internal.h:350]: i'd like to just have a single dispatcher class if possible. probably in another PR though. in the meantime, i'll add more to this comment

---
In reply to: [564864476](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564864476) [](ancestors = 564864476)

File: onnxruntime/core/providers/op_kernel_type_control.h (hunk 3 above)
Comment (edgchen1) [reply onnxruntime/core/providers/op_kernel_type_control.h:64]: good catch, that should be 'type' for OpKernelArgTypes

---
In reply to: [564874473](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564874473) [](ancestors = 564874473)

File: include/onnxruntime/core/framework/data_types_internal.h (hunk 2 above)
Comment (skottmckay) [reply include/onnxruntime/core/framework/data_types_internal.h:369]: ok. do we need to create an array though vs. just doing static_cast<void>(args) (which is what ORT_UNUSED_PARAMETER does)? 

---
In reply to: [564916970](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564916970) [](ancestors = 564916970,564866954)

File: onnxruntime/core/providers/op_kernel_type_control.h (hunk 3 above)
Comment (edgchen1) [reply onnxruntime/core/providers/op_kernel_type_control.h:64]: EnabledOpKernelArgTypes::type should always exist if compilation is successful. it ensures that OpKernelArgTypes<tags::Supported<OpKernelArgTag>> has a 'type' data member

---
In reply to: [564921386](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564921386) [](ancestors = 564921386,564874473)

File: onnxruntime/core/providers/op_kernel_type_control.h (hunk 4 above)
Comment (edgchen1) [reply onnxruntime/core/providers/op_kernel_type_control.h:69]: hm, a type list is also a type. :) i don't disagree and originally had 'types', but 'type' seems more conventional, at least with examples i've seen from std

---
In reply to: [564874953](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564874953) [](ancestors = 564874953)

File: include/onnxruntime/core/framework/data_types_internal.h (hunk 2 above)
Comment (edgchen1) [reply include/onnxruntime/core/framework/data_types_internal.h:369]: actually it was a warning for each arg within args. this array is one way to put ORT_UNUSED_PARAMETER on each arg.

---
In reply to: [564921484](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564921484) [](ancestors = 564921484,564916970,564866954)

File: onnxruntime/core/providers/cpu/tensor/cast_op.cc (hunk 7 above)
Comment (edgchen1) [reply onnxruntime/core/providers/cpu/tensor/cast_op.cc:106]: using unsigned long long here because of the call to std::stoull

---
In reply to: [564884415](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564884415) [](ancestors = 564884415)

File: onnxruntime/core/providers/cpu/tensor/cast_op.cc (hunk 8 above)
Comment (edgchen1) [reply onnxruntime/core/providers/cpu/tensor/cast_op.cc:32]: will try to do some measurements

---
In reply to: [564887790](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564887790) [](ancestors = 564887790)

File: onnxruntime/test/providers/op_kernel_type_control_test.cc (hunk 5 above)
Comment (edgchen1) [reply onnxruntime/test/providers/op_kernel_type_control_test.cc:33]: hm. i think it's a bit tricky. if we use a global list that actually disables types globally, then other tests may not work correctly. but if we can scope it to this translation unit (possibly just calling the macro in this file), then i think that's an ODR violation.

---
In reply to: [564880494](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564880494) [](ancestors = 564880494)

**Comment by edgchen1** [comment]
> 

we can update it, this is a placeholder.
an alternative is to pass in some arbitrary path to use as the type overrides file.

---
In reply to: [767869344](https://github.com/microsoft/onnxruntime/pull/6466#issuecomment-767869344) [](ancestors = 767869344)

---
Refers to: onnxruntime/core/providers/op_kernel_type_control_overrides.inc:11 in 882857a. [](commit_id = 882857a5671148e7bd9011f671a31c511e8e9ed6, deletion_comment = False)

File: onnxruntime/core/providers/op_kernel_type_control.h (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/providers/op_kernel_type_control.h:70]: I'd slightly prefer the readability of 'types' vs. having to know that 'type' is a type list given that's somewhat opaque. I would expect it's easier to comprehend what is going on with that naming if you're unfamiliar with the setup.

Are the examples in std dealing with a single type (e.g. type in a vector) or a type list?

Comment (edgchen1) [reply onnxruntime/core/providers/op_kernel_type_control.h:70]: ok. i haven't come across examples in std dealing with type lists.

---
In reply to: [564951914](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564951914) [](ancestors = 564951914)

File: onnxruntime/core/providers/op_kernel_type_control_overrides.inc [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+// Examples:
+// Specify allowed types per Op kernel arg:
+//   ORT_SPECIFY_OP_KERNEL_ARG_ALLOWED_TYPES(Cast, Input, 0, float, int64_t);
+//   ORT_SPECIFY_OP_KERNEL_ARG_ALLOWED_TYPES(Cast, Output, 0, float, int64_t);
+// Specify allowed types globally:
```
Comment (skottmckay) [inline onnxruntime/core/providers/op_kernel_type_control_overrides.inc:13]: Should this also include a domain in the tag? An operator name is only guaranteed to be unique within a domain afaik. 

e.g.
  ORT_SPECIFY_OP_KERNEL_ARG_ALLOWED_TYPES(ai.onnx, Cast, Input, 0, float, int64_t)

**Comment by skottmckay** [comment]
> 

nit: this file doesn't show up in VS (guessing we only look for .cc and .h files) so may need to manually add it to the project via cmake 

---
Refers to: onnxruntime/core/providers/op_kernel_type_control.h:211 in a2f9529. [](commit_id = a2f9529c08422b31874f259df1e0eeac2a5ab19a, deletion_comment = False)

File: onnxruntime/core/providers/op_kernel_type_control_overrides.inc (hunk 10 above)
Comment (edgchen1) [reply onnxruntime/core/providers/op_kernel_type_control_overrides.inc:49]: added an argument for specifying domain. i don't think it'll take ai.onnx, could we use something like kOnnxDomain?

---
In reply to: [565027440](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r565027440) [](ancestors = 565027440)

**Review by skottmckay — APPROVED** [review]

File: onnxruntime/core/providers/cpu/tensor/cast_op.cc (hunk 8 above)
Comment (edgchen1) [reply onnxruntime/core/providers/cpu/tensor/cast_op.cc:32]: ongoing, will do any necessary fixes in a future PR

---
In reply to: [564929687](https://github.com/microsoft/onnxruntime/pull/6466#discussion_r564929687) [](ancestors = 564929687,564887790)
